File: Tutorials/Canvas/Lesson005-Complex.ts
// This little trick (not to be copied!) is a temporary undocumented feature.
var log = function(s: any) {
  var w: any = window;
  w.Sk.output(s+'\n');
};

log('Hello, World!!');

var z = blade.Complex(2, 3);

log(z);

class Complex {
  public x: number;
  public y: number;
  constructor(x: number, y:number) {
    this.x = x;
    this.y = y;
  }
  arg(): number {
    return Math.atan2(this.y, this.x);
  }
  multiply(that: Complex): Complex {
    return new Complex(this.x * that.x - this.y * that.y, this.x * that.y + this.y * that.x);
  }
  toString(): String {
    return this.x + "+" + this.y + "i";
  }
}

var zero = new Complex(0,0);
var one = new Complex(1,0);
var i = new Complex(0,1);
var minusOne = new Complex(-1,0);
var minusI = new Complex(0,-1);

log("zero: " + zero);
log("one: " + one);
log("i: " + i);

log("arg(1) " + one.arg());
log("arg(i) " + i.arg());
log("arg(-1) " + minusOne.arg());
log("arg(-i) " + new Complex(0,-1).arg());

File: Tutorials/Canvas/Lesson005-Complex.py
class Complex:
  def __init__(self, x, y):
    self.x = x
    self.y = y

  def mod(self):
    return sqrt(self.x * self.x + self.y * self.y)

  def arg(self):
    return atan2(self.y, self.x)

  def __add__(self, other):
    return Complex(this.x + that.x, this.y + that.y)

  def __sub__(self, other):
    return Complex(this.x - that.x, this.y - that.y)

  def __mul__(self, other):
    return Complex(this.x * that.x - this.y * that.y, this.x * that.y + this.y * that.x)

  def __div__(self, other):
    denom = that.x * that.x + that.y * that.y;
    return Complex((this.x * that.x + this.y * that.y)/denom, (this.y * that.x - this.x * that.y)/denom)

  def str(self):
    return self.x + "+" + self.y + "i"

one = Complex(1.0,0.0)

print one